### ISSUE TITLE
Failure to push changes when files are in LFS

### ISSUE THREAD

**Issue by oschwald** [renovatebot/renovate#6842]
**What Renovate type are you using?**

Self-hosted with GitHub Enterprise

**Describe the bug**

We use Yarn's offline-mirror feature to vendor the packages. The vendored packages are stored in Git using LFS. When Renovate tries to push the changes, it fails with the message below.

**Relevant debug logs**

Slightly sanitized error message.

```
DEBUG: 35 file(s) to commit (repository=group/repo, branch=renovate/js)
DEBUG: Committing files to branch renovate/js (repository=group/repo, branch=renovate/js)
DEBUG: Error commiting files (repository=group/repo, branch=renovate/js)
       "err": {
         "task": {
           "commands": [
             "push",
             "origin",
             "renovate/js:renovate/js",
             "--force",
             "-u",
             "--no-verify",
             "--verbose",
             "--porcelain"
           ],
           "concatStdErr": true,
           "format": "utf-8"
         },
         "message": "Pushing to https://github.company.com/group/repo.git\nPOST git-receive-pack (12865 bytes)\nremote: error: GH008: Your push referenced at least 24 unknown Git LFS objects:        \nremote:     fa64832d250b8a4beea35965a3bb9ab26bddfeb73dbfe836ed61e70e3974fe18        \nremote:     efc11cec699211605376abf2a7aa47b7c7e1854d4428bcef9e67b258918d64d5        \nremote:     8e92658ac0ed2e085bbf11f21bb455d6a9eeae173de21d376e9bbd7105378f1e        \nremote:     ...        \nremote: Try to push them with 'git lfs push --all'.        \nerror: failed to push some refs to 'https://github.company.com/group/repo.git'\n", "stack": "Error: Pushing to https://github.company.com/group/repo.git\nPOST git-receive-pack (12865 bytes)\nremote: error: GH008: Your push referenced at least 24 unknown Git LFS objects:        \nremote:     fa64832d250b8a4beea35965a3bb9ab26bddfeb73dbfe836ed61e70e3974fe18        \nremote:     efc11cec699211605376abf2a7aa47b7c7e1854d4428bcef9e67b258918d64d5        \nremote:     8e92658ac0ed2e085bbf11f21bb455d6a9eeae173de21d376e9bbd7105378f1e        \nremote:     ...        \nremote: Try to push them with 'git lfs push --all'.        \nerror: failed to push some refs to 'https://github.company.com/group/repo.git'\n\n    at GitExecutorChain.onFatalException (/home/user/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.js:60:87)\n    at GitExecutorChain.<anonymous> (/home/user/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.js:51:28)\n    at Generator.throw (<anonymous>)\n    at rejected (/home/user/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.js:6:65)\n    at process._tickCallback (internal/process/next_tick.js:68:7)" }
DEBUG: Passing repository-changed error up (repository=group/repo, branch=renovate/js)
 INFO: Repository has changed during renovation - aborting (repository=group/repo)
```

**Comment by viceice** [comment]
For this we would need to make the renovate git layer git-lfs aware.

**Comment by viceice** [comment]
@oschwald Can you prepare a minimal public reproduction repo? So we have something to test against it.

**Comment by oschwald** [comment]
Thanks. I made an example repo at https://github.com/oschwald/renovate-lfs-issue.

With this repo, the exact failure is:

```
DEBUG: Error commiting files (repository=oschwald/renovate-lfs-issue, branch=renovate/js)
       "err": {
         "task": {
           "commands": [
             "push",
             "origin",
             "renovate/js:renovate/js",
             "--force",
             "-u",
             "--no-verify",
             "--verbose",
             "--porcelain"
           ],
           "concatStdErr": true,
           "format": "utf-8"
         },
         "message": "Pushing to https://github.com/oschwald/renovate-lfs-issue.git\nPOST git-receive-pack (4516 bytes)\nremote: error: GH008: Your push referenced at least 11 unknown Git LFS objects:        \nremote:     26acb6450ef03e44b1e1df2457588c0177013c9a6ded0f8e058c5d034a5391e2        \nremote:     47ab5c4571504bdee569f03e3423af5b51aa17d6a94866ddcae353ed2d9033eb        \nremote:     c609324ab889515f2f7354ddcc319b6080c9b76f2ac1441c03da031c85458696        \nremote:     ...        \nremote: Try to push them with 'git lfs push --all'.        \nerror: failed to push some refs to 'https://github.com/oschwald/renovate-lfs-issue.git'\n",
         "stack": "Error: Pushing to https://github.com/oschwald/renovate-lfs-issue.git\nPOST git-receive-pack (4516 bytes)\nremote: error: GH008: Your push referenced at least 11 unknown Git LFS objects:        \nremote:     26acb6450ef03e44b1e1df2457588c0177013c9a6ded0f8e058c5d034a5391e2        \nremote:     47ab5c4571504bdee569f03e3423af5b51aa17d6a94866ddcae353ed2d9033eb        \nremote:     c609324ab889515f2f7354ddcc319b6080c9b76f2ac1441c03da031c85458696        \nremote:     ...        \nremote: Try to push them with 'git lfs push --all'.        \nerror: failed to push some refs to 'https://github.com/oschwald/renovate-lfs-issue.git'\n\n    at GitExecutorChain.onFatalException (/home/greg/MaxMind/renovate-test/node_modules/simple-git/src/lib/runners/git-executor-chain.js:60:87)\n    at GitExecutorChain.<anonymous> (/home/greg/MaxMind/renovate-test/node_modules/simple-git/src/lib/runners/git-executor-chain.js:51:28)\n    at Generator.throw (<anonymous>)\n    at rejected (/home/greg/MaxMind/renovate-test/node_modules/simple-git/src/lib/runners/git-executor-chain.js:6:65)\n    at process._tickCallback (internal/process/next_tick.js:68:7)"
       }
```

**Comment by viceice** [comment]
I think this needs a new repo config options to configure files (minimatch / wildcard)  to checkout with git lfs.
By default git lfs should be disabled.

**Comment by sgtsquiggs** [comment]
git lfs uses a pre-push git hook, which is disabled by renovate using `--no-verify` #6521

**Comment by amacneil** [comment]
Hi friends,

What would it take to get git lfs pull support over the finish line? Can we contribute anything to help?

This is the one feature preventing us from using renovate, and like many others we are on yarn 2 with the yarn binary stored in lfs. 

Unfortunately, dependabot doesn’t support yarn 2, and renovate doesn’t support git lfs, so we’re in a bit of a rock and hard place.

**Comment by jtbandes** [comment]
Hi @rarkins – any chance you can comment on the above? 💕

**Comment by rarkins** [comment]
@fgblomqvist @viceice could either of you summarize the current status?

**Comment by viceice** [comment]
We still need to define if it will be always enabled? That way i assume it will always automatically checkout all lfs files. So we don't like to have that.

I think we need to add lfs install to buildpack without full enable it. So install binaries only.

Then we need a configurable list of lfs files, which renovate should checkout from  / commit to branch.

**Comment by amacneil** [comment]
Does anyone need this feature for anything other than yarn 2?

I think it would be sufficient for 99% of users to checkout a default list of files, e.g. `git lfs pull --include .yarn yarn.lock **/package-lock.json`.

If anyone arrives after that with a request for different files, then it could be made configurable (or just add to the lfs checkout list). I haven't heard any use case for making it configurable yet, but many people (like me) are stuck because of yarn 2.

**Comment by viceice** [comment]
Instead of fully disabling git lfs by default i like to now some common `lfs.fetchexclude` and /or `lfs.fetchinclude` settings for renovate to make this work for most users.

I plan to make those settings user configurable.

**Comment by amacneil** [comment]
Sounds great - looking forward to it!

**Comment by viceice** [comment]
So renovate defaults would be:
```js
{
  allowGitLfs: true,
  gitLfsInclude: '.yarn,yarn.lock,**/package-lock.json',
  gitLfsExclude: undefined,
}
```

`allowGitLfs: false` would override includes to `undefiend` and excludes to `*`.

Any other files to consider by default?

**Comment by amacneil** [comment]
I think those will cover our use case. 

Maybe call the flag `enableGitLfs: true` or even just `gitLfs: true` ?

**Comment by rarkins** [comment]
I'm all for letting self-hosted users get their hands on this first, and we work out how to handle it in the "untrusted" app environment later.

**Comment by amacneil** [comment]
Is it hard to get working in the shared environment? This doesn't seem like it should be any more or less safe than a regular git checkout.

**Comment by viceice** [comment]
@amacneil The hosted app has disabled git hooks for security concerns, so yes, it's not as easy to enable fls there.

**Comment by defunctzombie** [comment]
Is there still a plan to enable git lfs support for the hosted app? Or is the recommendation to self-host?

**Comment by viceice** [comment]
Yes, but still needs some time for all preparations.

**Comment by gjasny** [comment]
FYI: buildpack git-lfs install is broken: https://github.com/containerbase/buildpack/pull/238

**Comment by steschi** [comment]
What's the current status? What is blocking this? I think yarn-v3 and zero-installs with git-lfs are getting more common but without proper lfs support (at least opt-in or only on self-hosted) these repositories can not take advantage of renovate at all.

**Comment by AIGeneratedUsername** [comment]
It took some working hours for me to setup Renovate with git-lfs support (and a custom Python version), so I would like to share a working configuration with other people.

1. Build a custom Docker image. 
   ```dockerfile
   FROM renovate/renovate:34.24
   SHELL ["/bin/bash", "-o", "pipefail", "-c"]
   RUN install-tool python 3.10.8  \
       && install-tool poetry 1.2.2
   USER root
   RUN curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash  \
       && apt-get install -y --no-install-recommends git-lfs=3.2.0 \
       && apt-get -y autoremove && apt-get clean && rm -rf /var/lib/apt/lists/*
   USER 1000
   ```
   Note that I install a Python version that I use. If you are fine to use the latest Python (Python 3.11), then change this Dockerfile.
   Calling `USER root` is required before calling `apt-get` or the command will fail due to the lack of permissions.
2. Set required settings to support git-lfs.
   ```yml
   # This is mandatory for git-lfs or it will fail with the `GitLab: LFS objects are missing. Ensure LFS is properly set up or try a manual "git lfs push --all"`
   RENOVATE_GIT_NO_VERIFY: "[]"
   # Perhaps `RENOVATE_BINARY_SOURCE` is redundant if you already installed all required packages in your custom Docker image.
   RENOVATE_BINARY_SOURCE: "install"
   ```
   Other configuration options are omitted for the sake of brevity.

**Comment by jtbandes** [comment]
FYI, for those using yarn, dependabot now supports yarn berry and git lfs (https://github.com/dependabot/dependabot-core/pull/5833).

**Comment by rarkins** [comment]
@AIGeneratedUsername thanks for posting your solution. I have some questions on some of the things you added:

```
RUN install-tool python 3.10.8  \
    && install-tool poetry 1.2.2
```

I'm wondering why these are necessary. Shouldn't both of these be dynamically installed due to binarySource=install at runtime?

```
RENOVATE_BINARY_SOURCE: "install"
```

First of all, I'm not sure if this is *mandatory* for Git LFS in any way. But also, it is now the default behavior for `renovate/renovate` non-slim images anyway.

The main one of interest to me is:

```
RENOVATE_GIT_NO_VERIFY: "[]"
```

Here is the option for this:

```
  {
    name: 'gitNoVerify',
    description:
      'Which Git commands will be run with the `--no-verify` option.',
    type: 'array',
    subType: 'string',
    allowString: true,
    allowedValues: ['commit', 'push'],
    default: ['commit', 'push'],
    stage: 'global',
    globalOnly: true,
  },
```

So by setting this you're making sure Renovate runs e.g. `git push --no-verify`. Is this may be a workaround and not yet a solution?

**Comment by viceice** [comment]
disable git no-verify is required because LFS is working via git hooks

**Comment by rarkins** [comment]
Would we have any way in future to detect LFS is in use and remove `--no-verify` automatically?

**Comment by viceice** [comment]
not easily, we also set it to suppress arbitrary git hooks from repos for security reasons.

**Comment by mat813** [comment]
I started using [SheetJS](https://sheetjs.com/) in a projet, and [they recommend vendoring](https://docs.sheetjs.com/docs/getting-started/installation/nodejs) the package file because the npm registry has an outdated one, as it is a bit large, I used git lfs to store it, and now renovate cannot run yarn install, it errors out with:

> error Extracting tar content of undefined failed, the file appears to be corrupt: "Unexpected end of data"

**Comment by quine** [comment]
+1 for supporting Git LFS. We're using self-hosted Renovate, and we've got several Go repos with vendored modules (with files exceeding our GHE file size limit), thus requiring Git LFS.

**Comment by small-stone** [comment]
I have same problem:

WARN: Error updating branch (repository=ee-frontend, branch=branch-testfive/major-react-router-monorepo)
       "err": {
         "task": {
           "commands": ["checkout", "-f", "master", "--"],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "Your branch is up to date with 'origin/master'.\nAlready on 'master'\n\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting '.git/hooks/post-checkout'.\n\n",
         "stack": "Error: Your branch is up to date with 'origin/master'.\nAlready on 'master'\n\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting '.git/hooks/post-checkout'.\n\n\n    at Object.action (/opt/containerbase/tools/renovate/36.57.2/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/opt/containerbase/tools/renovate/36.57.2/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /opt/containerbase/tools/renovate/36.57.2/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/opt/containerbase/tools/renovate/36.57.2/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/opt/containerbase/tools/renovate/36.57.2/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/opt/containerbase/tools/renovate/36.57.2/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
       }
       
  
  
  **I urgently need support Git LFS**

**Comment by mat813** [comment]
Maybe it would be possible, for example, to start by adding git-lfs to the full image ?
That way it would be available, and people don't have to maintain custom renovate images.

**Comment by viceice** [comment]
simply install git-lfs on the fly before running renovate on your ci and configure `gitNoVerify`

https://github.com/containerbase/base/blob/6885c851db8d7f40aaea33b9c2fabc4652958e89/test/latest/Dockerfile#L162

https://docs.renovatebot.com/self-hosted-configuration/#gitnoverify

**Comment by ekristen** [comment]
I hit a problem with no git-lfs being install in the container with golang this week. 

TL;DR - I modified the docker run locally to test, install-tool followed by git lfs install --skip-repo (to install it for the user) made everything work as expected again.

https://github.com/renovatebot/renovate/discussions/25041#discussioncomment-7203684

**Comment by rarkins** [comment]
@viceice if we merely _install_ git-lfs in the images, will that cause any problems?

**Comment by ekristen** [comment]
I think you'd have to install the tool and also initialize it.

**Comment by ekristen** [comment]
```console
install-tool git-lfs v3.4.0 && git lfs install
```

**Comment by viceice** [comment]
> @viceice if we merely _install_ git-lfs in the images, will that cause any problems?

no

@ekristen the install tool command will already initialize 
https://github.com/containerbase/base/blob/80f6e73580b58d6ed7e9c708424124f3fbfb23b2/src/usr/local/containerbase/tools/v2/git-lfs.sh#L40-L44

**Comment by rarkins** [comment]
@viceice we'd need it installed in the slim, full and sidecar images, right?

**Comment by viceice** [comment]
not on sidecar.

we didn't yet, because there is no way for renovate to block it in the app

**Comment by cyrilc-pro** [comment]
FYI, one issue with Git LFS is that Renovate is running commands like `git show origin/master:package-lock.json`.
This command is OK as long as `package-lock.json` is not stored using Git LFS.

**Comment by cyrilc-pro** [comment]
@ekristen I'm OK with this being a limitation. In my case, I have a `package-lock.json` that is almost 1.5MB, which is why our company policy requires it to be stored with Git LFS.

@AIGeneratedUsername The symptom is that the `package-lock.json` is not updated along with the `package.json`. The Renovate logs contain a warning about the `package-lock.json` not being valid JSON:
```
WARN: Error parsing npm lock file (repository=<redacted>, branch=<redacted>-npm-minor)
      "err": {
        "message": "Unexpected token v in JSON at position 0",
        "stack": "SyntaxError: Unexpected token v in JSON at position 0\n    at JSON.parse (<anonymous>)\n    at parseLockFile (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/modules/manager/npm/utils.ts:10:27)\n    at writeExistingFiles (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/modules/manager/npm/post-update/index.ts:161:24)\n    at getAdditionalFiles (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/modules/manager/npm/post-update/index.ts:473:3)\n    at processBranch (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/workers/repository/update/branch/index.ts:471:31)\n    at writeUpdates (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/workers/repository/process/write.ts:159:17)\n    at update (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/workers/repository/process/extract-update.ts:214:11)\n    at Object.renovateRepository (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/workers/repository/index.ts:76:19)\n    at attributes.repository (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/workers/global/index.ts:184:11)\n    at start (/opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/containerbase/tools/renovate/37.56.3/node_modules/renovate/lib/renovate.ts:18:22"
      }
```
The ultimate result is the failure of the `npm install` command on the branch that Renovate created.

**Comment by ekristen** [comment]
At this point if package-lock.json or any file is in git-lfs the git show will not work, but as it stands right now, because they are in git-lfs, and since renovate doesn't support git-lfs, renovate shouldn't be working at all anyways.

I propose we get git-lfs support merged and then we can look to improve the `git show` command, theoretically you can do `git show origin/master:npm-package.json | git lfs smudge` 

So the code could be changed to do `git show` and look if version is present with git-lfs header and if so run it through git lfs smudge to get the actual file contents.

**Comment by rarkins** [comment]
I've lost track of what the next steps are

**Comment by viceice** [comment]
we don't support it because we blocked gut hooks for security reasons.

if you self-host, then you can configure to allow hooks and you need to install git-lfs

here's how to do in official renovate images:
https://github.com/containerbase/base/blob/821a5931113d622de322683957f0c435352f703c/test/latest/Dockerfile#L170

**Comment by ekristen** [comment]
@viceice so is this something that's not going to be supported in the hosted version?

**Comment by viceice** [comment]
- https://github.com/renovatebot/renovate/pull/10748

this was a starter to fully support lfs by renovate. feel free to take and complete it 🤗

**Comment by rarkins** [comment]
Not in the foreseeable future for security reasons

**Comment by viceice** [comment]
> @viceice so is this something that's not going to be supported in the hosted version? 

probably yes, until someone finds a way without git hooks.

**Comment by ekristen** [comment]
Thanks @rarkins and @viceice -- I'm pretty familiar with git-lfs but I went and did a bit of reading.  You said not going to be allowed due to `hooks` being not supported for security reasons, but as far as I know `git-lfs` doesn't actually use any hooks, at least not the traditional `.git/hooks`.

If we were to install `git-lfs` for the user running renovate and more specifically the `git` commands, the `git lfs install` simply adds a filter stanza to the users `~/.gitconfig` which is then used during clone and checkout to read the `.gitattributes` and look for matching filters to do special things, it's sort of hook-ish but not `.git/hooks`.

I've check my configuration, and I'm looking at some additional things right now but I can't find any use of hooks or reference to hooks. Am I missing something?

**Comment by viceice** [comment]
it's using hooks somewhere to push the lfs files, see comments above. that's why you can allow hooks for self-host

**Comment by ekristen** [comment]
@viceice ah, that's the missing link, for pushing. It's definitely not needed for pulling down. 

What if we supported git-lfs for cloning but not for pushing, make that a limitation. Then for cloning it would work for go mods and others that have hashes of their contents but wouldn't support renovating any files that are actually in git-lfs? 

Another approach could be to take the direnv.net approach and allow only hooks that meet a specific hash.  For example, we could allow the 3 hook files that get created so long as their hash matches. 

```bash
#!/bin/sh
command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting '.git/hooks/pre-push'.\n"; exit 2; }
git lfs pre-push "$@"
```

This is the `pre-push` in a cloned repo that has git-lfs file, renovate could allow this file to exist in the repo as long as the file matches `911a6ef2ce82426869f449da86b8ca5f7cb6205dfd8236b35ef277b26aef9fed` if it doesn't match it's removed? 

Thoughts?

**Comment by ekristen** [comment]
Really? OK. I don't want to muck up this issue, so I'll send an offline note. 

---

Would the solution of allowing git-lfs and the specific webhooks be acceptable if it could be hashed to be the known git-lfs hooks or would that be a non-starter as well.

**Comment by viceice** [comment]
@rarkins i think we need to revert git-lfs install. otherwise the hosted app will start downloading all lfs files on checkout.

we need a very defensive setting first: https://github.com/renovatebot/renovate/pull/10748

**Comment by vonsch** [comment]
Hello, may I ask if this somehow under progress and target v40 or v41 is realistic, please? We use self-hosted renovate and we are very interested in this feature because many of our git repos store upstream tarballs in git-lfs.

Maybe if there would be some possibility with git-lfs disabled by default, and we would be able to enable it let's say on our self-hosted renovate installation or on particular git repositories. We already tried to install `git-lfs` into renovate images, but then renovate started to fail with some .js errors/exceptions ...

**Comment by rarkins** [comment]
I would be happy if someone from the community were to do the work and figure out a way to introduce it in a secure and backwards-compatible manner.

**Comment by milas** [comment]
I have a Go _dependency_ that uses LFS, so by default, a Renovate run generates invalid checksums in `go.sum` because the LFS filter is not present.

I'm self-hosting Renovate in GitLab CI, so as a quick fix:
```yaml
before_script:
  - install-tool git-lfs 3.6.0
  - git lfs install
```

Be sure to use Renovate itself to track the Git LFS version: https://github.com/git-lfs/git-lfs/releases 😉 

Again, I'm using a self-hosted Renovate run against trusted repos, so this is fine for my use case and I want to share in case it helps someone else, but it does not address the core blockers.

**Comment by DellanX** [comment]
This pretty much has me blocked in my usage of renovate. It looks to me like renovate scrubs my git-lfs credentials, and no amount of trickery that I have applied can get them back.

Is there any workaround guidence, or any location for a discussion about how to implement support for lfs?
I am very unfamiliar with how this bots code operates, but I would be willing to assist in adding this functionality.

**Comment by rarkins** [comment]
Here's a massive report from ChatGPT Deep Research:

[Renovate Bot – Git LFS Support Recommendation.pdf](https://github.com/user-attachments/files/21808573/Renovate.Bot.Git.LFS.Support.Recommendation.pdf)

**Comment by rarkins** [comment]
Here's what I think is the summary:

Today, if a repo uses LFS but it's unimportant to Renovate (e.g. it's some JPEGs) then it makes no difference and Renovate works fine as it never needs to read or modify LFS blobs. However, if LFS blobs are needed by Renovate for either of the following reasons, it will be a problem:
- If LFS blobs are needed as part of extract then such dependencies are missed (read-only)
- If LFS blobs are needed by package managers as part of artifact updating then such updates may fail (read-only)
- If LFS blob are needed by Renovate ore package managers (read-write)

We could add git LFS and enable it system-wide in Renovate images without risk. Doing so would mean that git cloning would then replace pointers with blobs automatically and the two read-only cases would start working.

To support read-write means allowing git hooks on push, which in theory could be manipulated by a malicious user to point to their own hooks, so it may be a premium-only feature for Mend-hosted.

Step 1: adding and enabling LFS in the Renovate images by default
Step 2: more granular controls of hooks - bot admins can decide if they're allowed, repo admins can decide if they want them or not

**Comment by DellanX** [comment]
That report goes pretty well into some of the magic I was trying, to which I can see why my attempts failed. Here is my particular situation. But it all boils down to the fact that I am using yarn pnp for zero installs.

1. I have an app on a private repository that has many dependencies (hence using renovate to keep things tidy and up to date)
2. Some of my libraries are my own libraries hosted on a private npm server
3. I have a coworker who just does frontend stuff, whom doesn't need to have tokens to this private server (I don't mind giving him a token, but it adds overhead and complexity to setup that I'd prefer to avoid)
4. I use yarn pnp for zero installs to create a cache of zip files, which enables a zero install setup.
5. When renovate runs, it can read and modify the git-lfs files (the yarn cache of zip files), but it does not upload new ones correctly)

I do see now why lfs isn't added, it makes sense that the hooks are disabled. One thing that may be nice would be a configuration option for renovate to use a user specified git behavior. (I think I tried `localFs` to do this, but my experiments yielded that renovate was still performing a shallow clone). Atleast then a user could specify git to run with hooks at their own risk.

Or adding and enabling LFS and configuration to the hooks could work, I do see the hesitation there though

**Comment by viceice** [comment]
@rarkins to overcome expensive logs downloads I tried to implement lfs filters in renovate some time ago, but it caused a chicken and egg issue with cloning, reading renovate config, updating lfs filters and then pulling lfs files again.

if we simply install lfs, then you'll see a huge runtime increase for repos with big and / or many lfs files because we start downloading then automatically.

**Comment by ekristen** [comment]
What if the following approach was taken?

- Require renovate config to never be in LFS
- (Optional) Require no file renovate needs to commit to be LFS

1. Install LFS, but disable it by default on clone `GIT_LFS_SKIP_SMUDGE=1`
2. Check LFS configuration in renovate config
3. Apply any LFS things you want from config `(--include/--exclude)`
4. `git lfs pull` to fetch LFS files (if needed)

I think this should strike a balance. 

My use case is very basic, I have docs built into a golang codebase where images or other docs are LFS, without LFS renovate fails to generate proper hashes for the private go modules.

**Comment by DellanX** [comment]
I think that'd work. Then I could reassign the environment variable to disable skipping smudge. I'd need to experiment to know if it'd solve my problem.

**Comment by JosXa** [comment]
We're also on yarn zero installs where the generated zip files are checked in via LFS. As the entire toolchain runs in-house, enabling LFS would not pose that big a risk here. An option to "just enable it" would be amazing. The way it stands, Renovate won't run for us unfortunately.

**Comment by matdibu** [comment]
as a temporary imperfect workaround, I wrote a workflow that does:
1. checkout with `lfs: true`
1. `git lfs install`
1. `yarn install --no-immutable`
1. commit the git diff (only if it was a renovate-authored PR)

it allows CI to run normally, but renovate still complains about artifacts errors

**Comment by JosXa** [comment]
> 3\. ``
> 4\. commit the git diff (only if it was a renovate-authored PR)
> 
> it allows CI to run normally, but renovate still complains about artifacts errors

@matdibu You do this for branches Renovate has already generated, after the fact?

**Comment by matdibu** [comment]
yes, this runs on renovate-generated PRs, after they are opened

something like this:

```yaml
      - uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8  # v5.0.0
        with:
          ref: ${{ env.BRANCH }}
          token: ${{ secrets.GH_TOKEN_INFRA }}  # needed to commit and push under another user
          lfs: true

      - run: git lfs install

      - uses: ./.github/actions/setup-yarn

      - run: yarn install --no-immutable

      - name: Fail and open a PR with the diff
        run: |
          git diff --exit-code || (\
            echo '####################################################' && \
            echo '# a pull request will be generated with this diff  #' && \
            echo '# check your original PR for a mention             #' && \
            echo '# or just look for the last link in the step below #' && \
            echo '####################################################' && \
            exit 1)

      - uses: ./.github/actions/commit-or-pr-git-diff
        if: failure()
        with:
          token: ${{ secrets.GH_TOKEN_INFRA }}
          branch: ${{ env.BRANCH }}
          suffix: yarn
```

where BRANCH is defined as
```yaml
BRANCH: "${{ github.event.pull_request && github.head_ref || github.ref_name }}"
```
